# Malaria Study README
## 1. Data Collection
<p>The files for this study were downloaded from the BINP29 course server under /resources/binp29/Data/malaria and saved in the 0_Data folder. The entire analysis was run on the course server, unless stated otherwise.
Then, in a joint effort, the students of the course ran the gene prediction program GeneMark on it. 
In my case, I ran it on the genome of Plasmodium vivax.<p>

```shell
nohup gmes_petap.pl --ES --sequence plasmodium_vivax.genome &
```

<p>This produced a gtf file. The gtf files generated by the different students were collected in a temporary folder on the server and then downloaded into 1_GenePrediction_All, as each took a substantial amount of time to run.<p>

## 2. Processing of Haemoproteus tartakovskyi data
<p>The sequenced genome of Haemoproteus tartakovskyi in scaffolds could also be found on the server. But since malaria is a parasite, the reads do not only contain the parasites dna but also the hosts. We thus needed to filter the reads. This was done by GC content, as Haemoproteus tartakovskyi has a lower GC content (19-42%) compared to the host zebrafinch (41%). Furthermore, all reads shorter than 3000 nucleotides are sorted out as well.<p>

<p> This is done with the python script Scaffold.py.
The minimum scaffold length of 3000 was a given, the GC-content was chosen based on the lecture material. Run with<p>
  
```shell
python Scaffold.py Haemoproteus_tartakovskyi.genome 3000 35 Haemoproteus_genome_filtered.fasta
```
  
<p>The resulting fasta file contains 1681 reads, as opposed to the 2243 of the original. This high percentage is due to how the GC-content border was set. Since it is quite high, it is likely that more bird scaffolds are included but it also guarantees a majority of the Haemoproteus scaffolds to be included.<p>

<p> As a next step the outfile.fasta was also run through GeneMark for a gene prediction.<p>
  
 ```shell
 nohup gmes_petap.pl --ES --sequence Haemoproteus_genome_filtered.fasta & 
 ``` 
  
The resulting gtf file and the fasta produced earlier, together should create a new fasta file, using gffParse.pl (from the BINP29 courseserver). However, first we run into a issue with the headers of the fasta and the headers of the gtf file.                              
```shell
head -1 Haemoproteus.gtf
scaffold01477, GC: 24.2, Length: 3790	GeneMark.hmm	exon	21	1216	0+	.	gene_id "1_g"; transcript_id "1_t";
head -1 Haemoproteus_genome_filtered.fasta
>scaffold00001, GC: 27.63, Length: 116706
```
  
The fasta file recognizes spaces as columnseparators, whereas the gtf file considers tabs columnseperators. It therefore tries to match only the scaffold plus its number with the whole string. The solution is to remove the annotations from the id name in the gtf file. Furthermore, the commas in the fasta file headers are removed.
  
```shell
cat Haemoproteus.gtf | sed "s/,.*\tGeneMark.hmm/\tGeneMark.hmm/1" > Haemoproteus_altered.gtf
cat Haemoproteus_genome_filtered.fasta | tr -d ",">Haemoproteus_genome2_filtered.fasta
```
  
 So that the first line is now.
  
```shell
scaffold01477	GeneMark.hmm	exon	21	1216	0	+	.	gene_id "1_g"; transcript_id "1_t";
```
  
The files produced in the filtering and then geneprediction of Haemoproteus, are saved in 2_Filter_Pred_Haemoproteus. Using them, we can now run the gffParse.pl in a new folder 3_gffParse_Haemoproteus.

```shell
gffParse.pl -c -p -F -i ../2_Filter_Pred_Haemoproteus/Haemoproteus_genome2_filtered.fasta -g ../2_Filter_Pred_Haemoproteus/Haemoproteus_altered.gtf 
```
  
To further exclude scaffolds with bird origin, we use blastp (v. 2.11.0+) and the SwissProt database run in the folder 4_Blast_Haemoproteus.

```shell
blastp -query  ../gffParse_output/gffParse.faa -db SwissProt
```
 
From the output we filter out all the results that have their most significant hit in an avian species. For this we require the information from the NCBI taxonomy file (found on ftp://ftp.ebi.ac.uk/pub/databases/taxonomy/taxonomy.dat) and the Swissprot taxonomy file (ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz). In our case these files are already on the server, so instead we set up a symlink to them.

```shell
ln -s /resources/binp29/Data/malaria/taxonomy.dat taxonomy.dat
ln -s /resources/binp29/Data/malaria/uniprot_sprot.dat uniprot_sprot.dat
```
  
<p> The scaffolds are identified with the script BlastParser.py run in the command line with <p>
  
```shell
python BlastParser.py Blast/Ht_blastout.txt gffParse_output/gffParse.faa genome_files/Haemoproteus_genome2_filtered.fasta taxonomy.dat uniprot_sprot.dat Ht_without_birdscaff.fasta
```
<p>All the files resulting after removing the avian scaffolds were saved in the folder 5_remove_Birds.<p>
<p>Then the geneprediction done earlier for Haemoproteus is run again, because we do not want to predict bird genes. This is done in 6_Pred_Haemoproteus_without_bird.
  
```shell
nohup gmes_petap.pl --ES --sequence ../5_remove_Birds/Ht_without_birdscaff.fasta 
```
<p>To answer a few questions on size of the genomes, gene size and gc contents for all these organisms, we use bash. Please see the project report for the result table. <p>
```shell
#Genome size for all genome files.
for file in *.genome; do (echo $(grep -v "^>" $file | tr -d "\n" | wc -c)); done
#Number of genes for all gtf files.
for file in *.gtf; do cut -f9 $file | cut -d \" -f2 | sort | uniq | wc -l; done
#GC content for all genome files. Results for genome size, and counting G's and C's
for file in *.genome; do grep -v "^>" $file | tr -d "\n" | tr -d "N" | tr -d "A" | tr -d "T"| wc -c; done 
```
  
  
<p>Next we run the gffParse for all the genomes and their corresponding gtf files in folder 7_gffParse_All.<p>
```shell
gffParse.pl -c -p -i ../0_Data/Plasmodium_berghei.genome -g ../1_GenePrediction_All/plasmodium_berghei.gtf -b Pb
gffParse.pl -c -p -i ../0_Data/Plasmodium_cynomolgi.genome -g ../1_GenePrediction_All/plasmodium_cynomolgi.gtf -b Pc
gffParse.pl -c -p -i ../0_Data/Plasmodium_faciparum.genome -g ../1_GenePrediction_All/plasmodium_faciparum.gtf -b Pf
gffParse.pl -c -p -i ../0_Data/Plasmodium_knowlesi.genome -g ../1_GenePrediction_All/knowlesi.gtf -b Pk
gffParse.pl -c -p -i ../0_Data/Plasmodium_vivax.genome -g ../1_GenePrediction_All/plasmodium_vivax.gtf -b Pv
gffParse.pl -c -p -i ../0_Data/Plasmodium_yoelii.genome -g ../1_GenePrediction_All/plasmodium_yoelii.gtf -b Py
gffParse.pl -c -p -i ../0_Data/Toxoplasma_gondii.genome -g ../1_GenePrediction_All/Tg.gff -b Tg
```
  
  
  
  
  
  
